lib/cache/PackFileCacheStrategy.js

```diff
@@ -112,170 +124,448 @@ class Pack {
+	/**
+	 * Split large content files with used and unused items
+	 * into two parts to separate used from unused items
+	 */
+	_optimizeUnusedContent() {
+		// 1. Find a large content file with used and unused items
+		for (let i = 0; i < this.content.length; i++) {
+			const content = this.content[i];
+			if (content === undefined) continue;
+			const size = content.getSize();
+			if (size < MIN_CONTENT_SIZE) continue;
+			const used = content.used.size;
+			const total = content.items.size;
+			if (used > 0 && used < total) {
+				// 2. Remove this content
+				this.content[i] = undefined;
+
+				// 3. Determine items for the used content file
+				const usedItems = new Set(content.used);
+				const newLoc = this._findLocation();
+				this._gcAndUpdateLocation(usedItems, usedItems, newLoc);
+
+				// 4. Create content file for used items
+				if (usedItems.size > 0) {
+					this.content[newLoc] = new PackContent(
+						usedItems,
+						new Set(usedItems),
+						async () => {
+							await content.unpack();
+							const map = new Map();
+							for (const identifier of usedItems) {
+								map.set(identifier, content.content.get(identifier));
+							}
+							return map;
+						}
+					);
+				}
+
+				// 5. Determine items for the unused content file
+				const unusedItems = new Set(content.items);
+				const usedOfUnusedItems = new Set();
+				for (const identifier of usedItems) {
+					unusedItems.delete(identifier);
+				}
+				const newUnusedLoc = this._findLocation();
+				this._gcAndUpdateLocation(unusedItems, usedOfUnusedItems, newUnusedLoc);
+
+				// 6. Create content file for unused items
+				if (unusedItems.size > 0) {
+					this.content[newUnusedLoc] = new PackContent(
+						unusedItems,
+						usedOfUnusedItems,
+						async () => {
+							await content.unpack();
+							const map = new Map();
+							for (const identifier of unusedItems) {
+								map.set(identifier, content.content.get(identifier));
+							}
+							return map;
+						}
+					);
+				}
+
+				this.logger.log(
+					"Split pack %d into pack %d with %d used items and pack %d with %d unused items",
+					i,
+					newLoc,
+					usedItems.size,
+					newUnusedLoc,
+					unusedItems.size
+				);
+
+				// optimizing only one of them is good enough and
+				// reduces the amount of serialization needed
+				return;
 			}
 		}
-		this.lastAccess = read();
-		this.content = new Map();
-		let identifier = read();
-		while (identifier !== null) {
-			const entry = read();
-			if (typeof entry === "function") {
-				this.content.set(identifier, entry);
+	}
+
+	serialize({ write, writeSeparate }) {
+		this._persistFreshContent();
+		this._optimizeSmallContent();
+		this._optimizeUnusedContent();
+		for (const identifier of this.itemInfo.keys()) {
+			write(identifier);
+		}
+		write(null);
+		for (const info of this.itemInfo.values()) {
+			write(info.etag);
+		}
+		for (const info of this.itemInfo.values()) {
+			write(info.lastAccess);
+		}
```

Comment:
Sorry fo some nerding =) Could you explain value of writing in 3 cycles? As I understand it is optimized only for `BinaryMiddleware` (it allocates one section per `etags`, `lastAccess`, `identifier`'s).

Reply:
Yep, (de)serialization is more efficient, when similar values follow each other.

e. g. a run of numbers will be (de)serialized in one go with a single header -> https://github.com/webpack/webpack/blob/40074d308f784b0a736177b0aa5582a53e733345/lib/serialization/BinaryMiddleware.js#L385-L399
and a run of `null`s (most etags are null) -> https://github.com/webpack/webpack/blob/40074d308f784b0a736177b0aa5582a53e733345/lib/serialization/BinaryMiddleware.js#L430-L434
Maybe there will be a gzip middleware in future for network serialization that will also benefit from local similarity.

Similar pattern is found for other types too: e. g. Object and Map serializes all keys before all values. ReplaceSource splits replacements into numbers and strings.

I actually haven't measured if this really makes a difference. If you want to you can try to evaluate that.